File: 例子/Java的例子/pure/VerifySEStamp.java
import java.io.File;
import java.io.FileInputStream;
import net.netca.pki.encoding.asn1.pki.seseal.Stamp;
import net.netca.pki.SEStamp;
import net.netca.pki.encoding.Hex;
import net.netca.pki.encoding.asn1.pki.JCEVerifier;
import net.netca.pki.encoding.asn1.pki.seseal.Header;
import net.netca.pki.encoding.asn1.pki.seseal.StampInfo;
import net.netca.pki.encoding.asn1.pki.seseal.ESPropertyInfo;
import net.netca.pki.encoding.asn1.pki.seseal.ExtData;
import net.netca.pki.encoding.asn1.pki.seseal.ExtensionDatas;
import net.netca.pki.encoding.asn1.pki.seseal.CertDigestList;
import net.netca.pki.encoding.asn1.pki.seseal.CertDigestObj;
import net.netca.pki.encoding.asn1.pki.seseal.ESPictureInfo;
import net.netca.pki.encoding.asn1.pki.X509Certificate;


public class VerifySEStamp {

	public static void main(String[] args) throws Exception{
		if(args.length!=1) {
            System.out.println("Usage: java VerifySEStamp stampFileName");
            return;
        }
		String stampFileName=args[0];
		byte[] data=readFile(stampFileName);
		Stamp stamp=Stamp.decode(data);
		if(stamp.verifySignature(new JCEVerifier())==false) {
			System.out.println("verify fail");
		} else {
			System.out.println("verify ok");
		}
		
		int format=stamp.getStampType();
		if(format==Stamp.GBT_38540){
			System.out.println("GB/T 38540 stamp");
		} else if(format==Stamp.GMT_0031){
			System.out.println("GM/T 0031 stamp");
		} else {
			System.out.println("bad stamp format");
		}
		
		StampInfo info=stamp.getEstampInfo();
		Header header=info.getHeader();
		System.out.println("version:"+header.getVersion());
		System.out.println("vid:"+header.getVid());
		System.out.println("id:"+info.getEsID());
		
		ESPropertyInfo property=info.getProperty();
		System.out.println("type:"+property.getType());
		System.out.println("name:"+property.getName());
		
		int certListType=property.getCertListType();
		if(certListType==SEStamp.CERTLIST_TYPE_CERT){
			X509Certificate[] certs=property.getCertList();
			System.out.println("seal signer cert count:"+certs.length);
			for(int i=0;i<certs.length;i++) {
				System.out.println("\tcert "+i+":"+certs[i].getSubject().getLdapName());
			}
		} else {
			CertDigestList certDigestList=property.getCertDigestList();
			CertDigestObj certDigest;
			int count=certDigestList.size();
			System.out.println("seal signer cert digest count:"+certDigestList.size());
			for(int i=0;i<count;i++){
				certDigest=certDigestList.get(i);
				System.out.println("\tcert digest "+i+" type:"+certDigest.getType()+",cert digest "+i+" value:"+Hex.encode(true,certDigest.getValue()));
			}
		}
		System.out.println("createData:"+property.getCreateDate());
		System.out.println("validStart:"+property.getValidStart());
		System.out.println("validEnd:"+property.getValidEnd());
		
		ESPictureInfo picture=info.getPicture();
		System.out.println("picture type:"+picture.getType());
		System.out.println("picture width:"+picture.getWidth());
		System.out.println("picture height:"+picture.getHeight());
				
		ExtensionDatas exts=info.getExtDatas();
		if(exts==null) {
			System.out.println("no extension");
		} else {
			int count=exts.size();
			ExtData ext;
			System.out.println("extension count:"+count);
			for(int i=0;i<count;i++) {
				System.out.println("extension "+i);
				ext=exts.get(i);
				System.out.println("\textnID:"+ext.getOid());
				System.out.println("\tcritical:"+ext.isCritical());
				System.out.println("\textnValue:"+Hex.encode(true, ext.getExtensionValue()));
			}
		}
		
		System.out.println("stamp maker cert:"+stamp.getCert().getSubject().getLdapName());
		System.out.println("signAlgorithm:"+stamp.getSignatureAlgorithm());
		
	}
	private static byte[] readFile(String fileName) throws Exception{
		File file=new File(fileName);
		
		long fileLen=file.length();
		if(fileLen>=java.lang.Integer.MAX_VALUE) {
			throw new Exception("file too big:"+ fileName);
		}
		int len=(int)fileLen;
		byte[] data=new byte[len];
		FileInputStream fIn=new FileInputStream(file);
		
		try{
			if (fIn.read(data)!=len) {
				throw new Exception("read file fail:" + fileName);
			}
			return data;
		} finally {
			fIn.close();
		}
	}
}

File: 例子/lua的例子/VerifySEStamp.lua
local CryptCore=require("CryptCore")

function ToANSIString(utf8String)
    return CryptCore.UTF8ToACP(utf8String)
end


function Usage(filename)
    print(string.format("Usage: lua %s stampFileName",filename))
    print(ToANSIString("参数"))
    print(ToANSIString("\tstampFileName：电子印章的文件名"))
    print(ToANSIString("说明"))
    print(ToANSIString("\t验证stampFileName的电子印章"))
end

function GetFileConetent(fileName)
    local fIn=io.open(fileName,"rb")
    if fIn==nil then
        return nil
    end
    local data=fIn:read("*a")
    fIn:close()
    
    return data
end

if #arg~=1 then
    Usage(arg[0])
    os.exit(-1)
end

local stampFileName=arg[1]
local stampEncode=GetFileConetent(stampFileName)
if stampEncode==nil then
    print(ToANSIString("读电子印章文件"));
    os.exit(-1)
end

local stamp=CryptCore.NewStamp(stampEncode)
if stamp==nil then
    print(ToANSIString("不是电子印章文件失败"));
    os.exit(-1)
end

local format=stamp:GetFormat()
if format==CryptCore.Const["GB/T 38540"] then
    print(ToANSIString("符合GB/T 38540的电子印章"))
else
    print(ToANSIString("符合GM/T 0031的电子印章"))
end

print(ToANSIString("版本号：".. stamp:GetVersion()))
print(ToANSIString("Vid：".. stamp:GetVid()))
print(ToANSIString("Id：".. stamp:GetId()))
print(ToANSIString("类型：".. stamp:GetType()))
print(ToANSIString("名称：".. stamp:GetName()))


local certListType=stamp:GetCertListType()
if certListType==CryptCore.Const.STAMP_CERTLIST_TYPE_CERT then
    local certs=stamp:GetCertList()
    for i,cert in ipairs(certs) do
        print(ToANSIString("证书"..i..":" ..cert:GetSubject()))
        cert:Free()
    end
else
    local certDigestList=stamp:GetCertDigestList()
    for i,certDigest in ipairs(certDigestList) do
        print(ToANSIString("证书"..i.."的杂凑类型:"..certDigest.type..",杂凑值:" ..CryptCore.HexEncode(certDigest.value)))
    end
end

print(ToANSIString("创建时间：".. stamp:GetCreateDate()))
print(ToANSIString("有效期开始时间：".. stamp:GetValidStart()))
print(ToANSIString("有效期结束时间：".. stamp:GetValidEnd()))

local pictureType,pictureData,pictureWidth,pictureHeigth=stamp:GetPicture()
print(ToANSIString("图片类型：".. pictureType))
print(ToANSIString("图片宽度：".. pictureWidth))
print(ToANSIString("图片高度：".. pictureHeigth))


local exts=stamp:GetExtension()
if #exts~=0 then
    local oid
    local critical
    local extValue
    for i,ext in ipairs(exts) do
        oid=ext[1]
        critical=ext[2]
        extValue=ext[3]
        if critical then
            print(ToANSIString(string.format("关键扩展(%s)：%s", oid,CryptCore.HexEncode(extValue))))
        else
            print(ToANSIString(string.format("非关键扩展(%s)：%s", oid,CryptCore.HexEncode(extValue))))
        end
    end
else
    print(ToANSIString("没有扩展"))
end

local signCert=stamp:GetSignCert()
print(ToANSIString("制章人证书：".. signCert:GetSubject()))
signCert:Free()

print(ToANSIString("签名算法：".. stamp:GetSignAlgorithm()))

if stamp:Verify() then
    print(ToANSIString("验证电子印章成功"))
else
    print(ToANSIString("验证电子印章失败"))
end

stamp:Free()
